### PR TITLE
Send/receive ECN bits on linux

### PR DESCRIPTION
**Pull request by Ralith**
Not sure we want to merge this before libc merges my PR there, but there's lots of esoterica to scrutinize in the meantime.

### CODE REVIEW

**Comment by djc** [comment]
Can you link to your libc PR from here? Also I think it'd be good to discuss at least with the tokio-udp folks if they want to take this stuff before we land it here?

**Comment by Ralith** [comment]
See https://github.com/rust-lang/libc/pull/1175.

I'm not sure what integration into tokio-udp would look like; ECN is just one of many(?) possible types of ancillary data that can be sent/received alongside packets for niche purposes. A sufficiently generic API seems likely to be deemed out of scope, as it would probably dwarf the other parts of the crate. This also doesn't strike me as a maintenance hazard; udp.rs is the full extent of the code approximately duplicated from tokio-udp, and is trivial.

We may also want to digress further from the beaten bath API in the future. For example, we can use the same interfaces on Linux to get access to hardware timestamps for high-precision timing of packet receipt times for improved congestion control, and we might want to adopt `sendmmsg`/`recvmmsg` for reduced context switching. Using our own UDP binding will make this easier.

That said, I suppose it won't hurt to kick off some discussion and see where it goes.

**Comment by Ralith** [comment]
Opened https://github.com/tokio-rs/tokio/issues/799.

File: quinn/src/platform/linux.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    io, mem,
+    net::{SocketAddr, SocketAddrV4, SocketAddrV6},
+    ptr,
+};
+
+use mio::net::UdpSocket;
+
+use quinn_proto::EcnCodepoint;
+
+const CMSG_LEN: usize = 24;
```
Comment (djc) [inline quinn/src/platform/linux.rs]: Nit: put this at the bottom?

File: quinn/src/platform/fallback.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+use mio::net::UdpSocket;
+
+use quinn_proto::EcnCodepoint;
+
+pub fn init(socket: &UdpSocket) -> io::Result<()> {}
```
Comment (djc) [inline quinn/src/platform/fallback.rs]: Feels like this wants to be trait.

File: quinn/src/platform/fallback.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+use std::{io, net::SocketAddr};
+
+use mio::net::UdpSocket;
+
+use quinn_proto::EcnCodepoint;
+
+#[allow(dead_code)]
+pub struct Platform;
```
Comment (djc) [inline quinn/src/platform/fallback.rs]: Why do we still need this?

**Comment by djc** [comment]
Approved, though I would be in favor of squashing these commits before merging.

**Pull request merged**